warehouse-management-service/internal/handler/handler.go

```diff
@@ -469,7 +479,226 @@ func (h *handler) DeleteShelfBlock(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 	}
+}
+
+func (h *handler) GetShelf(w http.ResponseWriter, r *http.Request) {
+	shelfId := chi.URLParam(r, "shelfId")
+
+	if shelfId == "" {
+		err := fmt.Errorf("%v", api.ShelfResponse{
+			Error: "shelf id cannot be empty",
+		})
+		h.logger.Log(log.Error, err)
+		h.response(w, http.StatusBadRequest, err)
+		return
+	}
 
+	shelf, err := h.shelfService.GetShelfById(r.Context(), shelfId)
+	switch err {
+	case wms.ShelfDoesNotExist:
+		{
+			h.logger.Log(log.Error, err)
+			h.response(w, http.StatusNotFound, api.ShelfResponse{Error: fmt.Sprintf(
+				"failed to get, shelf: %s does not exist",
+				shelfId,
+			)})
+		}
+	case nil:
+		{
+			h.response(w, http.StatusOK, api.ShelfResponse{Response: shelf})
+		}
+	default:
+		{
+			h.logger.Log(log.Error, err)
+			h.response(w, http.StatusInternalServerError, api.ShelfResponse{Error: "Failed to get shelf"})
+		}
+	}
+}
+
+func (h *handler) CreateShelf(w http.ResponseWriter, r *http.Request) {
+	var createShelfRequest api.CreateShelfRequest
+
+	if r.Body == nil {
+		err := fmt.Errorf("request body cannot be empty")
+		h.logger.Log(log.Error, err)
+		h.response(w, http.StatusBadRequest, api.ShelfResponse{
+			Error: err.Error()})
+		return
+	}
+
+	decoder := json.NewDecoder(r.Body)
+	decoder.DisallowUnknownFields()
+	err := decoder.Decode(&createShelfRequest)
+	if err != nil {
+		h.logger.Log(log.Error, err)
+		h.response(w, http.StatusBadRequest, api.ShelfResponse{
+			Error: "Failed to parse request"})
+		return
+	}
+
+	err = validator.Validate(createShelfRequest)
+	if err != nil {
+		h.logger.Log(log.Error, err)
+		h.response(w, http.StatusBadRequest, api.ShelfResponse{
+			Error: fmt.Sprintf("Invalid input: %v", err.Error())})
+		return
+	}
+
+	shelf := wms.NewShelf(
```

Comment:
I think `CreateShelfRequest`, `UpdateShelfRequest` & `Shelf` can all be clubbed into 1 struct. It's ok to share a model-level struct at API level. If not very varying, 1 struct can have multiple tags too for JSON & SQL

Reply:
Updated this to use `Shelf` in the handler as well